### ISSUE TITLE
Following number keeps growing but there's no one to unfollow

### ISSUE THREAD

**Issue by martinkrulltott** [InstaPy/InstaPy#4645]
**TL;DR; Accounts that were followed but never unfollowed are missing from my followedPool.csv, how do I find these rogue followed accounts?**

I'm running InstaPy twice a day and it's set to do three things - _like_by_tags_, _follow_user_followers_ and _unfollow_users_ (in this particular order). I rarely follow users manually but to prevent unfollowing of these users it's set to `InstapyFollowed=(True, "all"), style="FIFO", unfollow_after=72*60*60`. I.e. unfollow after 3 days.
Sometimes my quota supervisor kicks in for daily server calls, which makes the last action - _unfollow_users_ stop. 

I'm expecting the amount of followed to stay roughly the same over a long period of time, but I've noticed that it's been steadily increasing by a few hundred over the last few months. 

Since the quota supervisor sometimes cuts the _unfollow_users_ short, I was expecting there to be a "backlog" of users that it hasn't had time to unfollow yet, however checking the logs I can see multiple occasions of _Total 0 users available to unfollow_. I can also see that the followedPool.csv only contains users from the last 3 days, meaning that InstaPy indeed thinks that there's nobody to unfollow, even though there are hundreds of accounts that were followed by InstaPy that were never unfollowed.

To prevent my followed list to keep growing I need to find these rogue followed accounts to unfollow. Instagram lists followed accounts in an arbitrary order, so I can't use that to figure out which were followed by me and which were "lost" by InstaPy. However I have access to _record_all_followed.csv_, so I should be able to compare these accounts to my current followed accounts and create a _customList_ to unfollow. However my python skills aren't that great.

1. How do you create a list of _record_all_followed.csv_ accounts matched with "my current followed accounts" to be used as a custom unfollow list?

2. How can I prevent InstaPy from losing followed accounts in the future?

_ps my whole bot.py script is pretty long and hard to follow, since it's using a lot of variables and dynamic behaviour, which is why I didn't include it here_

**Comment by yoannes** [comment]
Hi,

I have the same problem. Many times Instapy says that unfollowed someone but it didn't.. 

Since I don't want to touch the Instapy class, because I pull the changes from github.

So I made my own follow/unfollow method, I keep a list of everyone I followed with timestamp and unfollow when I want to. However, even after asking Instapy to unfollow, in the end I fetch all my following to insure that each user was really unfollowed..

That was the only quick solution for me.

```
            obj = load_obj() # load followed from file
            to_unfollow = []
            for user in obj["followed"]:
                x = time.time() - obj["followed"][user]
                if x > unfollow_after:  # time in seconds to unfollow
                    to_unfollow.append(user)
            to_unfollow = to_unfollow[0:acc["total_unfollow"]] # limit to unfollow

            if len(to_unfollow):
                session.unfollow_users(amount=acc["total_unfollow"],
                                                     customList=(True, to_unfollow, "all"),
                                                     style="RANDOM", unfollow_after=1, sleep_delay=120)

                following = session.grab_following(username=acc["username"],
                                                                        amount="full",
                                                                        live_match=True,
                                                                        store_locally=False)
                for user in list(obj["followed"]):
                    if user not in following:
                        del obj["followed"][user]

                dump_obj(obj) # save obj back to file
```

**Comment by martinkrulltott** [comment]
Thanks for the answer @yoannes.
I put together my own solution based on _grab_following_, the _record_all_followed_ csv and _unfollow_users_ .
Basically it's grabbing all the current followed users, matching them to the _all followed users_ from the csv and unfollowing the ones that are older than 5 days (to let the unfollow method of my regular bot handle the more recent accounts). 
I've tried it a bit and it seems to be working to find the rogue accounts that are still followed but aren't detected by InstaPy.

        import json
        import csv
        from datetime import datetime
        from datetime import timedelta
        from instapy import InstaPy
        from instapy.util import smart_run
        from random import randint
        from random import sample
        import sys
        insta_username = 'YOURACCOUNT'
        insta_password = 'YOURPASSWORD'
        cutOffDate = datetime.today() - timedelta(days=5)
        following = []
        unfollowAmount = 50

        def readMyFile(filename):
            dates = []
            accounts = []
            with open(filename) as csvDataFile:
                csvReader = csv.reader(csvDataFile, delimiter='~')
                for row in csvReader:
                    dates.append(row[0].strip())
                    accounts.append(row[1].strip())
                
            return dates, accounts

        print("Starting bot with unfollow amount " + str(unfollowAmount) + ", cut off date " + str(cutOffDate))

        session = InstaPy(username=insta_username,
                        password=insta_password,
                        headless_browser=True,
                        nogui=True)

        with smart_run(session):
            try:
                # following = json.load(open('logs/' + insta_username + '/relationship_data/' + insta_username + '/following/NAMEOFJSONFILE.json', "r"))
                following = session.grab_following(username=insta_username, amount="full", live_match=False, store_locally=True)
                print("Current follow amount: " + str(len(following)))
            except Exception as e:
                print("ERROR with grab_following: " + str(e))

            followedDates,followedAccounts = readMyFile('logs/' + insta_username + '/' + insta_username + '_record_all_followed.csv')

            rogue = []

            for account in following:
                index = followedAccounts.index(account) if account in followedAccounts else -1
                if (index > -1):
                    date = datetime.strptime(followedDates[index], '%Y-%m-%d %H:%M')
                    print(followedAccounts[index] + ': ' + str(date) + ' < ' + str(cutOffDate) + ' ' + str(date < cutOffDate))
                    if (date < cutOffDate):
                        rogue.append(followedAccounts[index])
            print("Current rogue amount: " + str(len(rogue)))
            try:
                session.unfollow_users(amount=unfollowAmount, customList=(True, rogue, "all"), unfollow_after=None, sleep_delay=randint(600, 700))
            except Exception as e:
                print("ERROR with unfollow_users: " + str(e))

**Comment by KeenonT** [comment]
Super helpful. Everything was working fine up until two days ago and the unfollowing function stopped working normally.

Martin, would you still share your script? I'd be interested to see what you've added in terms of dynamic behavior ?

**Comment by yoannes** [comment]
thats nice. Good that is working
